File: app/src/main/java/com/seafile/seadroid2/framework/model/sdoc/FileRecordWrapperModel.java
package com.seafile.seadroid2.framework.model.sdoc;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class FileRecordWrapperModel implements Parcelable {
    public List<MetadataModel> metadata = new ArrayList<>();
    public List<Map<String, Object>> results = new ArrayList<>();

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeTypedList(this.metadata);
        dest.writeList(this.results);
    }

    public FileRecordWrapperModel() {
    }

    protected FileRecordWrapperModel(Parcel in) {
        this.metadata = in.createTypedArrayList(MetadataModel.CREATOR);
        this.results = new ArrayList<Map<String, Object>>();
        in.readList(this.results, Map.class.getClassLoader());
    }

    public static final Creator<FileRecordWrapperModel> CREATOR = new Creator<FileRecordWrapperModel>() {
        @Override
        public FileRecordWrapperModel createFromParcel(Parcel source) {
            return new FileRecordWrapperModel(source);
        }

        @Override
        public FileRecordWrapperModel[] newArray(int size) {
            return new FileRecordWrapperModel[size];
        }
    };
}

File: app/src/main/java/com/seafile/seadroid2/framework/model/sdoc/FileRecordWrapperModel.kt
package com.seafile.seadroid2.framework.model.sdoc

import android.os.Parcel
import android.os.Parcelable

class FileRecordWrapperModel() : Parcelable {
    @JvmField
    var metadata: MutableList<MetadataModel> = ArrayList()

    @JvmField
    var results: MutableList<Map<String, Any?>> = ArrayList()

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeTypedList(metadata)
        dest.writeList(results)
    }

    private constructor(parcel: Parcel) : this() {
        metadata = parcel.createTypedArrayList(MetadataModel.CREATOR) ?: ArrayList()
        val list = ArrayList<Map<String, Any?>>()
        parcel.readList(list, Map::class.java.classLoader)
        results = list
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<FileRecordWrapperModel> =
            object : Parcelable.Creator<FileRecordWrapperModel> {
                override fun createFromParcel(source: Parcel): FileRecordWrapperModel = FileRecordWrapperModel(source)
                override fun newArray(size: Int): Array<FileRecordWrapperModel?> = arrayOfNulls(size)
            }
    }
}
